### PR TITLE
Clean up datastore creator and get storeid from page config

### PR DESCRIPTION
**Pull request by saulshanabrook**
Builds on https://github.com/jupyterlab/jupyterlab/pull/6871

- [x] Remove extra `TableManager` abstraction. We moved the `changed` signal to the `TableManager` and will move the typing of the schema to the core `Datastore` object in the future. Ideally, it would be nice to remove the `TableManager` abstraction as well, eventually.
- [x] Put the storeid on the page config, instead of sending a request for it.
- [x] Try removing datastore manager, making it all async
- [x] Fix empty history responses

Now this depends on https://github.com/jupyterlab/jupyterlab_server/pull/74

### CODE REVIEW

File: packages/datastore/src/token.ts [hunk 1]
```diff
@@ -3,66 +3,18 @@
...
-    this._currentDatastore = this.manager.datastore;
-    this._currentDatastore.changed.connect(this._datastoreSlot, this);
-  }
-
-  private _currentDatastore: Datastore;
-  private readonly _changed: Signal<unknown, unknown>;
-}
 export class DatastoreCreator {
-  async createTable<S extends Schema>(
+  async createTable(
```
Comment (saulshanabrook) [inline packages/datastore/src/token.ts]: should be `createDatastore`

File: jupyterlab/datastore/handler.py [hunk 2]
```diff
@@ -96,150 +85,130 @@ def get(self, *args, **kwargs):
...
-    def pre_get(self):
-        # authenticate first
-        super(CollaborationHandler, self).pre_get()
-
-        self.store_id = self.get_query_argument('storeId', None)
-        # TODO: Check if store id supplied, if so, check if already open
-        if self.store_id is not None:
-            # TODO: Temporary minimal reaction:
-            self.log.warning("Trying to reopen store that is already connected")
-            raise web.HTTPError(400)
```
Comment (vidartf) [inline jupyterlab/datastore/handler.py]: The bit before this acts as a pretty decent placeholder (raises if trying to connect multiple times with the same store id).

Comment (vidartf) [reply jupyterlab/datastore/handler.py]: We should probably also fail here if the `store_id` is not given.

File: packages/datastore/src/client.ts [hunk 3]
```diff
@@ -172,11 +158,9 @@ export class CollaborationClient extends WSConnection<
     if (token) {
       queryParams.push(`token=${encodeURIComponent(token)}`);
     }
-    if (this._storeId !== null) {
-      queryParams.push(
-        `storeId=${encodeURIComponent(this._storeId.toString(10))}`
-      );
-    }
+    // queryParams.push(
```
Comment (ian-r-rose) [inline packages/datastore/src/client.ts]: Is this intended?

File: packages/datastore/src/client.ts [hunk 4]
```diff
@@ -229,18 +213,21 @@ export class CollaborationClient extends WSConnection<
...
-        MessageLoop.postMessage(this.handler, message);
+    if (!this.handler) {
+      return;
+    }
+    for (let t of transactions) {
+      if (t.serial !== this._serverSerial + 1) {
+        // Out of order serials!
+        // Something has gone wrong somewhere.
+        // TODO: Trigger recovery?
+        throw new Error('Critical! Out of order transactions in datastore.');
```
Comment (ian-r-rose) [inline packages/datastore/src/client.ts]: @vidartf Did we ever decide whether this was actually true? It's my impression that things can be out-of-order at the transaction level as well as the patch level.

File: packages/datastore/src/manager.ts [hunk 5]
```diff
@@ -1,211 +1,34 @@
...
-    ...overrides,
-    id: newId,
-    schemas: toArray(map(source.iter(), table => table.schema))
+import { PageConfig } from '@jupyterlab/coreutils';
+
+export async function createDatastore(
+  collaborationId: string,
+  schemas: ReadonlyArray<Schema>
+): Promise<Datastore> {
+  const client = new CollaborationClient({
```
Comment (ian-r-rose) [inline packages/datastore/src/manager.ts]: This looks good to me for a first pass at this!

File: packages/datastore/src/manager.ts [hunk 6]
```diff
@@ -1,211 +1,34 @@
...
-    this
-  );
-}
-
-export namespace DatastoreManager {
-  export interface IChangedArgs {
-    datastore: Datastore;
-  }
-}
+export type SchemaFields = {
```
Comment (ian-r-rose) [inline packages/datastore/src/manager.ts]: I've killed this type in the other PR.

File: packages/datastore/src/wsmessages.ts [hunk 7]
```diff
@@ -38,8 +38,6 @@ export namespace Collaboration {
    * Maps a msgType to the corresponding type.
    */
   export interface ITypeMap {
-    'storeid-request': StoreIdRequest;
```
Comment (ian-r-rose) [inline packages/datastore/src/wsmessages.ts]: Eventually we may want to reconsider this, but I think having the storeid on page config is a good place for us to be right now.

File: packages/datastore/src/client.ts (hunk 4 above)
Comment (vidartf) [reply packages/datastore/src/client.ts]: Yeah, this was based on an incorrect assumption about which algorithm was being used (triggered by missing graveyard implementation). The entire serial setup needs a pass. It is probably still needed for other things, but this exception will likely go away.

I can look at it later hopefully.

File: packages/datastore/src/wsmessages.ts (hunk 7 above)
Comment (vidartf) [reply packages/datastore/src/wsmessages.ts]: It's in the git history if somebody ever thinks to resurrect it. Until then it is dead code, so it should go away :)

Comment (ian-r-rose) [reply packages/datastore/src/wsmessages.ts]: I am very much in favor of getting rid of dead code.

File: packages/datastore/src/client.ts (hunk 3 above)
Comment (saulshanabrook) [reply packages/datastore/src/client.ts]: I heard you don't like dead code :) I'll delete this.

File: packages/datastore/src/client.ts (hunk 4 above)
Comment (saulshanabrook) [reply packages/datastore/src/client.ts]: Yeah I didn't fully grok the serial interface. I tried to leave mainly as is for now.

File: packages/datastore/src/manager.ts (hunk 6 above)
Comment (saulshanabrook) [reply packages/datastore/src/manager.ts]: Sounds good.

File: packages/datastore/src/wsmessages.ts (hunk 7 above)
Comment (saulshanabrook) [reply packages/datastore/src/wsmessages.ts]: Cool, so is this good or are you suggesting a change?

Comment (ian-r-rose) [reply packages/datastore/src/wsmessages.ts]: Yeah, this is good.

File: jupyterlab/datastore/handler.py (hunk 2 above)
Comment (saulshanabrook) [reply jupyterlab/datastore/handler.py]: I removed the backend knowledge of the `store_id` altogether. I actually think this is  maybe ok? It just identifies users by connection.  If you lose connection, then you make a new one and it treats you as a new peer.

Comment (vidartf) [reply jupyterlab/datastore/handler.py]: It possibly makes things like checkpoints a little harder. I'll make a separate issue for checkpoints. We can reintroduce this concept if needed after that.

**Pull request merged**